---
original title: "plotgardener"
new title: occur_plot
author: "Jessica Honorato Mauer"
date: "3/12/2022 - 4/10/2022"
adapted by: "Liriel Almodobar"
date:
output: html_document
---

## Installing packages and databases

```{r, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install(version = "3.17")

#Transcriptome ranges for the known gene track of Homo sapiens, e.g., introns, exons, UTR regions. usar o do seu genoma de ref!!
BiocManager::install("TxDb.Hsapiens.UCSC.hg38.knownGene")

#Gene-based information for Homo sapiens; useful for mapping between gene IDs, Names, Symbols, GO and KEGG identifiers, etc.

BiocManager::install("org.Hs.eg.db")

#Provides a convenient interface to annotations from many different sources; objects are returned as fully parsed Bioconductor data objects or as the name of a file on disk.

BiocManager::install("AnnotationHub")

#plotgardener package

BiocManager::install("plotgardener")

```

```{r, echo=FALSE}
## Load libraries and datasets
library("plotgardener")
library("org.Hs.eg.db")
library("TxDb.Hsapiens.UCSC.hg38.knownGene")
library("AnnotationHub")
```

### Formatting miscall tables

Alterar o valor do cromossomo e rodar os demais comandos dessa seÃ§Ã£o
```{r}
i = 1
```


```{r}
# State and ancestry specific occurence counts
## All ancestries
read.table("/home/yuri/liri/puzzle/109_preliminar_oct23/infos_txt/sort_pos_count_chr_1_all.txt", h=F) -> rs_count

rs_count <- rs_count[,c(2,3,4)]
colnames(rs_count) <- c("chrom","bp", "score")
paste("chr",rs_count$chrom,sep='') -> rs_count$chrom

#scale(rs_count$score) -> rs_count$wc_scaled
#as.numeric(rs_count$wc_scaled) -> rs_count$wc_scaled

rs_count$end <- c(rs_count$bp[2:length(rs_count$bp)] - 1, rs_count$bp[length(rs_count$bp)] + 1)

rs_count <- rs_count[,c(1,2,5,3)]
colnames(rs_count) <- c("chrom", "start", "end", "score")

```

```{r}

## EUR
read.table("/home/yuri/liri/puzzle/109_preliminar_oct23/sort_pos_count_chr_1_eur.txt", h=F) -> count_eur_rs

count_eur_rs <- count_eur_rs[,c(2,3,4)]
colnames(count_eur_rs) <- c("chrom","bp", "score")
paste("chr",count_eur_rs$chrom,sep='') -> count_eur_rs$chrom

#scale(count_eur_rs$score) -> count_eur_rs$wc_scaled
#as.numeric(count_eur_rs$wc_scaled) -> count_eur_rs$wc_scaled

count_eur_rs$end <- c(count_eur_rs$bp[2:length(count_eur_rs$bp)] - 1, count_eur_rs$bp[length(count_eur_rs$bp)] + 1)

count_eur_rs <- count_eur_rs[,c(1,2,5,3)]
colnames(count_eur_rs) <- c("chrom", "start", "end", "score")

## AFR
read.table("/home/yuri/liri/puzzle/109_preliminar_oct23/sort_pos_count_chr_1_afr.txt", h=F) -> count_afr_rs

count_afr_rs <- count_afr_rs[,c(2,3,4)]
colnames(count_afr_rs) <- c("chrom","bp", "score")
paste("chr",count_afr_rs$chrom,sep='') -> count_afr_rs$chrom

#scale(count_afr_rs$score) -> count_afr_rs$wc_scaled
#as.numeric(count_afr_rs$wc_scaled) -> count_afr_rs$wc_scaled

count_afr_rs$end <- c(count_afr_rs$bp[2:length(count_afr_rs$bp)] - 1, count_afr_rs$bp[length(count_afr_rs$bp)] + 1)

count_afr_rs <- count_afr_rs[,c(1,2,5,3)]
colnames(count_afr_rs) <- c("chrom", "start", "end", "score")

## NAT
read.table("/home/yuri/liri/puzzle/109_preliminar_oct23/sort_pos_count_chr_1_nat.txt", h=F) -> count_nat_rs

count_nat_rs <- count_nat_rs[,c(2,3,4)]
colnames(count_nat_rs) <- c("chrom","bp", "score")
paste("chr",count_nat_rs$chrom,sep='') -> count_nat_rs$chrom

#scale(count_nat_rs$score) -> count_nat_rs$wc_scaled
#as.numeric(count_nat_rs$wc_scaled) -> count_nat_rs$wc_scaled

count_nat_rs$end <- c(count_nat_rs$bp[2:length(count_nat_rs$bp)] - 1, count_nat_rs$bp[length(count_nat_rs$bp)] + 1)

count_nat_rs <- count_nat_rs[,c(1,2,5,3)]
colnames(count_nat_rs) <- c("chrom", "start", "end", "score")

## UNK
read.table("/home/yuri/liri/puzzle/109_preliminar_oct23/sort_pos_count_chr_1_unk.txt", h=F) -> count_unk_rs

count_unk_rs <- count_unk_rs[,c(2,3,4)]
colnames(count_unk_rs) <- c("chrom","bp", "score")
paste("chr",count_unk_rs$chrom,sep='') -> count_unk_rs$chrom

#scale(count_unk_rs$score) -> count_unk_rs$wc_scaled
#as.numeric(count_unk_rs$wc_scaled) -> count_unk_rs$wc_scaled

count_unk_rs$end <- c(count_unk_rs$bp[2:length(count_unk_rs$bp)] - 1, count_unk_rs$bp[length(count_unk_rs$bp)] + 1)

count_unk_rs <- count_unk_rs[,c(1,2,5,3)]
colnames(count_unk_rs) <- c("chrom", "start", "end", "score")

```


### plotting parameters

```{r}
max_score <- max(c(rs_count$score, count_nat_rs$score, count_unk_rs$score, count_eur_rs$score, count_afr_rs$score)) #inclui global aqui?

print(max_score)

params <- pgParams(
    chrom = "chr1", chromstart = 1, chromend = 248956422,
    assembly = "hg38",
    x = 0.5, just = c("left", "top"),
    width = 4, length = 4, default.units = "inches",
    range = c(0, max_score)
)

```


## Quick plot signal data = if i only wanna look at a single ancxstate combination (below: total occurences for rs)
```{r}
pageCreate(
    width = 5, height = 2, default.units = "inches", xgrid = 0, ygrid = 0
)

ideogramPlot <- plotIdeogram(params= params,  
    y = 0.5,height = 0.2,
)

signalplot <- plotSignal(
    data = rs_count,
    linecolor = "darkgray",
    params = params,
    y = "0.4b", height = 0.5,
)

plotGenomeLabel(
    params = params,
    y = "0.01b", height = 0.2, scale = "Mb",
)


## Annotate y-axis
annoYaxis(
    plot = signalplot, at = as.character(seq(0, max_score+50, by = 50)),label=as.character(seq(0, max_score+50, by = 50)),
    axisLine = TRUE, fontsize = 9
)

## Plot y-axis label
plotText(
    label = "variants occurrences", x = 0.10, y = 1.37, rot = 90,
    fontsize = 9, fontface = "plain", just = "center",
    default.units = "inches"
)

pageGuideHide()

```

## direcao do erro v2
```{r}
pageCreate(
    width = 5, height = 6, default.units = "inches", xgrid = 0, ygrid = 0
)

ideogramPlot <- plotIdeogram(params= params,  
    y = 0.5,height = 0.2,
)

signalplot <- plotSignal(
    data = rs_count,
    linecolor = "darkgray",
    params = params,
    y = "0.4b", height = 0.5,
)

## Annotate y-axis
annoYaxis(
    plot = signalplot, at = seq(0, max_score+50, by = 50),label=as.character(seq(0, max_score+50, by = 50)),#acho que aqui ficava so at ou label, e tem que rever esse max_score+50
    axisLine = TRUE, fontsize = 9 #n lembro se mudava fonte, mas acho que sim
)

## Plot y-axis label
plotText(
    label = "variants occurrences", x = 0.10, y = 1.37, rot = 90, #adjust x, y and fontsize
    fontsize = 9, fontface = "plain", just = "center",
    default.units = "inches"
)

## Plot and place signal plots

signalplotEUR <- plotSignal(
    data = eur_rs_count,
    linecolor = "#7297b4",
    params = params,
    y = "0.03b", height = 0.5,
)

## Annotate y-axis
annoYaxis(
    plot = signalplot, at = as.character(seq(0, max_score+50, by = 50)),label=as.character(seq(0, max_score+50, by = 50)),
    axisLine = TRUE, fontsize = 9
)

signalplotAFR <- plotSignal(
    data = count_afr_rs,
    linecolor = "#c05040",
    params = params,
    y = "0.03b", height = 0.5,
)

## Annotate y-axis
annoYaxis(
    plot = signalplot, at = as.character(seq(0, max_score+50, by = 50)),label=as.character(seq(0, max_score+50, by = 50)),
    axisLine = TRUE, fontsize = 9
)

signalplotNAT <- plotSignal(
    data = nat_rs_count,
    linecolor = "#6caf5e",
    params = params,
    y = "0.03b", height = 0.5,
)

## Annotate y-axis
annoYaxis(
    plot = signalplot, at = as.character(seq(0, max_score+50, by = 50)),label=as.character(seq(0, max_score+50, by = 50)),
    axisLine = TRUE, fontsize = 9
)

signalplotUNK <- plotSignal(
    data = unk_rs_count,
    linecolor = "#919191",
    params = params,
    y = "0.03b", height = 0.5,
)

## Annotate y-axis
annoYaxis(
    plot = signalplot, at = as.character(seq(0, max_score+50, by = 50)),label=as.character(seq(0, max_score+50, by = 50)),
    axisLine = TRUE, fontsize = 9
)

plotGenomeLabel(
    params = params,
    y = "0.01b", height = 0.2, scale = "Mb",
)

#ALL NAT EUR AFR UNK
legendPlot <- plotLegend(
    legend = c("ALL","EUR","AFR","NAT","UNK"),
    fill = c("black","#7297b4","#c05040", "#919191"), #when in rstudio, replace by the colors and change in chr_frags
    border = FALSE,
    x = 5, y = 2.5, width = 1.5, height = 2,
    just = c("left", "top"),
    default.units = "inches"
)


pageGuideHide()

```
