---
output:
  html_document: default
  pdf_document: default
created_by: Jessica Mauer
adapted_by: Liriel Almodobar
---
## Installing packages and databases

```{r, eval=FALSE, echo = FALSE, message = FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install

#Transcriptome ranges for the known gene track of Homo sapiens, e.g., introns, exons, UTR regions. usar o do seu genoma de ref!!
BiocManager::install("TxDb.Hsapiens.UCSC.hg38.knownGene")

#Gene-based information for Homo sapiens; useful for mapping between gene IDs, Names, Symbols, GO and KEGG identifiers, etc.

BiocManager::install("org.Hs.eg.db")

#Provides a convenient interface to annotations from many different sources; objects are returned as fully parsed Bioconductor data objects or as the name of a file on disk.

BiocManager::install("AnnotationHub")

#plotgardener package

BiocManager::install("plotgardener")

```

```{r, echo=FALSE, message = FALSE}
## Load libraries and datasets
library("plotgardener")
library("org.Hs.eg.db")
library("TxDb.Hsapiens.UCSC.hg38.knownGene")
library("AnnotationHub")

```

## Formatting occurrence tables

Chromosome number
```{r}
i = 1
```

```{r}

## NAT RS
read.table("/home/yuri/liri/puzzle/rs/nat/chr_info_unfilt/count_info/count_chr_1_nat_rs.txt", h=F) -> count_nat_rs
count_nat_rs <- count_nat_rs[order(count_nat_rs[,3]),]
count_nat_rs <- count_nat_rs[,c(2,3,4)]
colnames(count_nat_rs) <- c("chrom","bp", "score")
paste("chr",count_nat_rs$chrom,sep='') -> count_nat_rs$chrom
count_nat_rs$end <- c(count_nat_rs$bp[2:length(count_nat_rs$bp)] - 1, count_nat_rs$bp[length(count_nat_rs$bp)] + 1)

count_nat_rs <- count_nat_rs[,c(1,2,4,3)]
colnames(count_nat_rs) <- c("chrom", "start", "end", "score")
#count_nat_rs$score <- count_nat_rs$score - 339 ## version with negative values

##NAT SP

read.table("/home/yuri/liri/puzzle/sp/nat/chr_info_unfilt/count_info/count_chr_1_nat_sp.txt", h=F) -> count_nat_sp
count_nat_sp <- count_nat_sp[order(count_nat_sp[,3]),]
count_nat_sp <- count_nat_sp[,c(2,3,4)]
colnames(count_nat_sp) <- c("chrom","bp", "score")
paste("chr",count_nat_sp$chrom,sep='') -> count_nat_sp$chrom
count_nat_sp$end <- c(count_nat_sp$bp[2:length(count_nat_sp$bp)] - 1, count_nat_sp$bp[length(count_nat_sp$bp)] + 1)

count_nat_sp <- count_nat_sp[,c(1,2,4,3)]
colnames(count_nat_sp) <- c("chrom", "start", "end", "score")
count_nat_sp$score <- count_nat_sp$score + 104 
#count_nat_sp$score <- count_nat_sp$score - 339 ##version with negative values

```


## Plotting parameters

```{r}
max_score <- max(count_nat_rs$score, count_nat_sp$score)
#min_score <- min(count_nat_rs$score, count_nat_sp$score) ## version with negative values
print(max_score)
#print(min_score) ## version with negative values

params <- pgParams(
    chrom = "chr1", 
    chromstart = 1, chromend = 249000000, 
    assembly = "hg38",
    x = 0.5, just = c("left", "top"),
    width = 4, length = 4, default.units = "inches",
    range = c(0, max_score)
    #range = c(min_score, max_score) ## version with negative values
)

```
## Obtaining signal plot and gene information
All regions
```{r}
pageCreate(
    width = 6, height = 5, default.units = "inches", xgrid = 0, ygrid = 0
)

## Plot and place ideogram

ideogramPlot <- plotIdeogram(params= params,
    y = 0.5, height = 0.2,
)

## Plot and place signal plots

## Plot
signalplot <- plotSignal(
    data = count_nat_rs,
    linecolor = "darkgreen",
    params = params,
    y = "0.2b", height = 1.0,
    #negData = TRUE ##version with negative values
)

### Annotate y-axis
annoYaxis(
    plot = signalplot, at = seq(0, max_score+50, by = 50),
    #plot = signalplot, at = seq(min_score, max_score+10, by = 10) ##version with negative values
    axisLine = FALSE, fontsize = 9
)

### Plot

signalplotNATSP <- plotSignal(
    data = count_nat_sp,
    linecolor = "green",
    params = params,
    y = "0.2b", height = 1.0,
    #negData = TRUE ##version with negative values
)

### Annotate y-axis
annoYaxis(
    plot = signalplotNATSP, at = seq(0, max_score+50, by = 50),
    #plot = signalplot, at = seq(min_score, max_score+10, by = 10) ##version with negative values
    axisLine = FALSE, fontsize = 9
)

### Plot y-axis label
plotText(
    label = "variants occurrences", x = 0.005, y = 1.90, rot = 90,
    fontsize = 9, fontface = "plain", just = "center",
    default.units = "inches"
)

### Annotate X axis
annoXaxis(
    plot = signalplotNATSP,
    at = format(seq(1, 249000000, by = 20000000),scientific = TRUE),
    axisLine = TRUE, fontsize = 9,
)

plotText(
  label = "chr1",
  x = 2.5,
  y = "0.43b",
  fontsize = 9
)

### Plot legend
legendPlot <- plotLegend(
    legend = c("NAT_RS","NAT_SP"),
    fill = c("darkgreen", "green"),
    border = FALSE,
    x = 4.5, y = 2.40, width = 1.5, height = 1.0,
    just = c("left", "top"),
    default.units = "inches"
)

# We can then use annoHighlight() to highlight our genomic region of interest with a box of our desired height:
region <- pgParams(chrom = "chr1", chromstart = 13000000, chromend = 24000000)
annoHighlight(
    plot = ideogramPlot, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)
region <- pgParams(chrom = "chr1", chromstart = 60000000, chromend = 76000000)
annoHighlight(
    plot = signalplotNATSP, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)

region <- pgParams(chrom = "chr1", chromstart = 166000000, chromend = 176000000)
annoHighlight(
    plot = ideogramPlot, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)
region <- pgParams(chrom = "chr1", chromstart = 220000000, chromend = 250000000)
annoHighlight(
    plot = ideogramPlot, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)

pageGuideHide()

```
Region 1
```{r}
pageCreate(
    width = 6, height = 5, default.units = "inches", xgrid = 0, ygrid = 0
)

ideogramPlot <- plotIdeogram(params= params,
    y = 0.5, height = 0.2,
)

signalplot <- plotSignal(
    data = count_nat_rs,
    linecolor = "darkgreen",
    params = params,
    y = "0.2b", height = 1.0,
)

## Annotate y-axis
annoYaxis(
    plot = signalplot, at = seq(0, max_score+50, by = 50),label=as.character(seq(0, max_score+50, by = 50)),#acho que aqui ficava so at ou label, e tem que rever esse max_score+50
    axisLine = FALSE, fontsize = 9
)

## Plot y-axis label
plotText(
    label = "variants occurrences", x = 0.005, y = 1.90, rot = 90,
    fontsize = 9, fontface = "plain", just = "center",
    default.units = "inches"
)

## Plot and place signal plots

signalplotNATSP <- plotSignal(
    data = count_nat_sp,
    linecolor = "green",
    params = params,
    y = "0.2b", height = 1.0,
)

## Annotate y-axis
annoYaxis(
    plot = signalplotNATSP, at = as.character(seq(0, max_score+50, by = 50)),label=as.character(seq(0, max_score+50, by = 50)),
    axisLine = FALSE, fontsize = 9
)

annoXaxis(
    plot = signalplotNATSP, at = format(seq(1, 249000000, by = 20000000),scientific = TRUE),
    axisLine = TRUE, fontsize = 9, 
)


plotText(
  label = "chr1",
  x = 2.5,
  y = "0.43b",
  fontsize = 9
)
# NATRS NATSP
legendPlot <- plotLegend(
    legend = c("NAT_RS","NAT_SP"),
    fill = c("darkgreen", "green"),
    border = FALSE,
    x = 4.5, y = 2.40, width = 1.5, height = 1.0,
    just = c("left", "top"),
    default.units = "inches"
)

# We can then use annoHighlight() to highlight our genomic region of interest with a box of our desired height:

region <- pgParams(chrom = "chr1", chromstart = 13000000, chromend = 24000000)
annoHighlight(
    plot = ideogramPlot, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)
#To make it clearer that we are zooming in on a genomic region, we can then use annoZoomLines() to add zoom lines from the genomic region we highlighted:

annoZoomLines(
    plot = ideogramPlot, params = region,
    y0 = 0.75, x1 = c(0.6, 4.5), y1 = 1.9, default.units = "inches"
)

#Finally, we can add our zoomed-in gene data within the zoom lines (indicate same reg as highlight):
plotGenes(params = region, assembly = "hg38",
   x = 0.6, y = 4, width = 4, height = 1, just = "left"
)

## Plot genome label
plotGenomeLabel(params = region,
    assembly = "hg38",
    x = 0.6, y = 4.5, length = 4,
    default.units = "inches"
)
region <- pgParams(chrom = "chr1", chromstart = 60000000, chromend = 76000000)
annoHighlight(
    plot = signalplotNATSP, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)
region <- pgParams(chrom = "chr1", chromstart = 166000000, chromend = 176000000)
annoHighlight(
    plot = ideogramPlot, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)
region <- pgParams(chrom = "chr1", chromstart = 220000000, chromend = 250000000)
annoHighlight(
    plot = ideogramPlot, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)

pageGuideHide()

```
Region 2
```{r}
pageCreate(
    width = 6, height = 5, default.units = "inches", xgrid = 0, ygrid = 0
)

ideogramPlot <- plotIdeogram(params= params,
    y = 0.5, height = 0.2,
)

signalplot <- plotSignal(
    data = count_nat_rs,
    linecolor = "darkgreen",
    params = params,
    y = "0.2b", height = 1.0,
)

## Annotate y-axis
annoYaxis(
    plot = signalplot, at = seq(0, max_score+50, by = 50),label=as.character(seq(0, max_score+50, by = 50)),#acho que aqui ficava so at ou label, e tem que rever esse max_score+50
    axisLine = FALSE, fontsize = 9
)

## Plot y-axis label
plotText(
    label = "variants occurrences", x = 0.005, y = 1.90, rot = 90,
    fontsize = 9, fontface = "plain", just = "center",
    default.units = "inches"
)

## Plot and place signal plots

signalplotNATSP <- plotSignal(
    data = count_nat_sp,
    linecolor = "green",
    params = params,
    y = "0.2b", height = 1.0,
)

## Annotate y-axis
annoYaxis(
    plot = signalplotNATSP, at = as.character(seq(0, max_score+50, by = 50)),label=as.character(seq(0, max_score+50, by = 50)),
    axisLine = FALSE, fontsize = 9
)

annoXaxis(
    plot = signalplotNATSP, at = format(seq(1, 249000000, by = 20000000),scientific = TRUE),
    axisLine = TRUE, fontsize = 9, 
)

plotText(
  label = "chr1",
  x = 2.5,
  y = "0.43b",
  fontsize = 9
)
# NATRS NATSP
legendPlot <- plotLegend(
    legend = c("NAT_RS","NAT_SP"),
    fill = c("darkgreen", "green"),
    border = FALSE,
    x = 4.5, y = 2.40, width = 1.5, height = 1.0,
    just = c("left", "top"),
    default.units = "inches"
)

# We can then use annoHighlight() to highlight our genomic region of interest with a box of our desired height:

region <- pgParams(chrom = "chr1", chromstart = 60000000, chromend = 76000000)
annoHighlight(
    plot = signalplotNATSP, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)

#To make it clearer that we are zooming in on a genomic region, we can then use annoZoomLines() to add zoom lines from the genomic region we highlighted:

annoZoomLines(
    plot = ideogramPlot, params = region,
    y0 = 0.75, x1 = c(0.6, 4.5), y1 = 1.9, default.units = "inches"
)

#Finally, we can add our zoomed-in gene data within the zoom lines (indicate same reg as highlight):
plotGenes(params = region, assembly = "hg38",
   x = 0.6, y = 4, width = 4, height = 1, just = "left"
)

## Plot genome label
plotGenomeLabel(params = region,
    assembly = "hg38",
    x = 0.6, y = 4.5, length = 4,
    default.units = "inches"
)
region <- pgParams(chrom = "chr1", chromstart = 13000000, chromend = 24000000)
annoHighlight(
    plot = ideogramPlot, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)
region <- pgParams(chrom = "chr1", chromstart = 166000000, chromend = 176000000)
annoHighlight(
    plot = ideogramPlot, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)
region <- pgParams(chrom = "chr1", chromstart = 220000000, chromend = 250000000)
annoHighlight(
    plot = ideogramPlot, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)


pageGuideHide()

```
Region 3
```{r}
pageCreate(
    width = 6, height = 5, default.units = "inches", xgrid = 0, ygrid = 0
)

ideogramPlot <- plotIdeogram(params= params,
    y = 0.5, height = 0.2,
)

signalplot <- plotSignal(
    data = count_nat_rs,
    linecolor = "darkgreen",
    params = params,
    y = "0.2b", height = 1.0,
)

## Annotate y-axis
annoYaxis(
    plot = signalplot, at = seq(0, max_score+50, by = 50),label=as.character(seq(0, max_score+50, by = 50)),#acho que aqui ficava so at ou label, e tem que rever esse max_score+50
    axisLine = FALSE, fontsize = 9
)

## Plot y-axis label
plotText(
    label = "variants occurrences", x = 0.005, y = 1.90, rot = 90,
    fontsize = 9, fontface = "plain", just = "center",
    default.units = "inches"
)

## Plot and place signal plots

signalplotNATSP <- plotSignal(
    data = count_nat_sp,
    linecolor = "green",
    params = params,
    y = "0.2b", height = 1.0,
)

## Annotate y-axis
annoYaxis(
    plot = signalplotNATSP, at = as.character(seq(0, max_score+50, by = 50)),label=as.character(seq(0, max_score+50, by = 50)),
    axisLine = FALSE, fontsize = 9
)

annoXaxis(
    plot = signalplotNATSP, at = format(seq(1, 249000000, by = 20000000),scientific = TRUE),
    axisLine = TRUE, fontsize = 9, 
)

plotText(
  label = "chr1",
  x = 2.5,
  y = "0.43b",
  fontsize = 9
)
# NATRS NATSP
legendPlot <- plotLegend(
    legend = c("NAT_RS","NAT_SP"),
    fill = c("darkgreen", "green"),
    border = FALSE,
    x = 4.5, y = 2.40, width = 1.5, height = 1.0,
    just = c("left", "top"),
    default.units = "inches"
)

# We can then use annoHighlight() to highlight our genomic region of interest with a box of our desired height:

region <- pgParams(chrom = "chr1", chromstart = 166000000, chromend = 176000000)
annoHighlight(
    plot = ideogramPlot, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)

#To make it clearer that we are zooming in on a genomic region, we can then use annoZoomLines() to add zoom lines from the genomic region we highlighted:

annoZoomLines(
    plot = ideogramPlot, params = region,
    y0 = 0.75, x1 = c(0.6, 4.5), y1 = 1.9, default.units = "inches"
)

#Finally, we can add our zoomed-in gene data within the zoom lines (indicate same reg as highlight):
plotGenes(params = region, assembly = "hg38",
   x = 0.6, y = 4, width = 4, height = 1, just = "left"
)

## Plot genome label
plotGenomeLabel(params = region,
    assembly = "hg38",
    x = 0.6, y = 4.5, length = 4,
    default.units = "inches"
)
region <- pgParams(chrom = "chr1", chromstart = 13000000, chromend = 24000000)
annoHighlight(
    plot = ideogramPlot, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)
region <- pgParams(chrom = "chr1", chromstart = 60000000, chromend = 76000000)
annoHighlight(
    plot = signalplotNATSP, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)

region <- pgParams(chrom = "chr1", chromstart = 220000000, chromend = 250000000)
annoHighlight(
    plot = ideogramPlot, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)

pageGuideHide()

```
Region 4
```{r}
pageCreate(
    width = 6, height = 5, default.units = "inches", xgrid = 0, ygrid = 0
)

ideogramPlot <- plotIdeogram(params= params,
    y = 0.5, height = 0.2,
)

signalplot <- plotSignal(
    data = count_nat_rs,
    linecolor = "darkgreen",
    params = params,
    y = "0.2b", height = 1.0,
)

## Annotate y-axis
annoYaxis(
    plot = signalplot, at = seq(0, max_score+50, by = 50),label=as.character(seq(0, max_score+50, by = 50)),#acho que aqui ficava so at ou label, e tem que rever esse max_score+50
    axisLine = FALSE, fontsize = 9
)

## Plot y-axis label
plotText(
    label = "variants occurrences", x = 0.005, y = 1.90, rot = 90,
    fontsize = 9, fontface = "plain", just = "center",
    default.units = "inches"
)

## Plot and place signal plots

signalplotNATSP <- plotSignal(
    data = count_nat_sp,
    linecolor = "green",
    params = params,
    y = "0.2b", height = 1.0,
)

## Annotate y-axis
annoYaxis(
    plot = signalplotNATSP, at = as.character(seq(0, max_score+50, by = 50)),label=as.character(seq(0, max_score+50, by = 50)),
    axisLine = FALSE, fontsize = 9
)

annoXaxis(
    plot = signalplotNATSP, at = format(seq(1, 249000000, by = 20000000),scientific = TRUE),
    axisLine = TRUE, fontsize = 9, 
)

plotText(
  label = "chr1",
  x = 2.5,
  y = "0.43b",
  fontsize = 9
)
# NATRS NATSP
legendPlot <- plotLegend(
    legend = c("NAT_RS","NAT_SP"),
    fill = c("darkgreen", "green"),
    border = FALSE,
    x = 4.5, y = 2.40, width = 1.5, height = 1.0,
    just = c("left", "top"),
    default.units = "inches"
)

# We can then use annoHighlight() to highlight our genomic region of interest with a box of our desired height:

region <- pgParams(chrom = "chr1", chromstart = 220000000, chromend = 249000000)
annoHighlight(
    plot = ideogramPlot, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)
#To make it clearer that we are zooming in on a genomic region, we can then use annoZoomLines() to add zoom lines from the genomic region we highlighted:

annoZoomLines(
    plot = ideogramPlot, params = region,
    y0 = 0.75, x1 = c(0.6, 4.5), y1 = 1.9, default.units = "inches"
)

#Finally, we can add our zoomed-in gene data within the zoom lines (indicate same reg as highlight):
plotGenes(params = region, assembly = "hg38",
   x = 0.6, y = 4, width = 4, height = 1, just = "left"
)

## Plot genome label
plotGenomeLabel(params = region,
    assembly = "hg38",
    x = 0.6, y = 4.5, length = 4,
    default.units = "inches"
)
region <- pgParams(chrom = "chr1", chromstart = 60000000, chromend = 76000000)
annoHighlight(
    plot = signalplotNATSP, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)
region <- pgParams(chrom = "chr1", chromstart = 13000000, chromend = 24000000)
annoHighlight(
    plot = ideogramPlot, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)
region <- pgParams(chrom = "chr1", chromstart = 166000000, chromend = 176000000)
annoHighlight(
    plot = ideogramPlot, params = region,
    fill = "salmon",
    y = 0.45, height = 0.3, just = c("left", "top"), default.units = "inches", alpha = 0.3
)


pageGuideHide()

```
